feat: Adding Utils Directory and new workflow file for Self-Updating Readme

LukaFontanilla opened the pull request:
<!--
Thank you for proposing a pull request! Please note that SOME TESTS WILL
LIKELY FAIL due to how GitHub exposes secrets in Pull Requests from forks.
Someone from the team will review your Pull Request and respond.

Please describe your change and any implementation details below.
-->



This PR is meant to automate self updating the deploy-cloud-function readme to include up to date event trigger types from gcloud. Example output below:

<img width="1126" alt="Screen Shot 2022-03-22 at 3 15 16 PM" src="https://user-images.githubusercontent.com/51471651/159588036-d5bce940-b2a0-406f-9f4c-56439c05b73f.png">


This PR specifically adds:

- `Utils` directory for self-updating readme files, which include (`main.moustache`: readme template file with variables returned from gcloud command.  `updateReadme.ts`: contains script to grab updated event trigger types from gcloud command, then writes new readme file. `tsconfig.json`: self-contained to utils directory, used to build update readme script)
- `updateReadme.yml` contains the workflow file that uses our auth action, builds and runs update readme script, and uses peter evans' create pull request action of a specific version - `peter-evans/create-pull-request@dcd5fd746d53dd8de555c0f10bca6c35628be47a` to commit Readme from machine user fork.
- `build-readme` new script in package.json that builds `updateReadme.ts` and outputs to a tmp folder

Other notes:
- Due to self contained tsconfig.json in the utils folder, excluding `updateReadme.ts` from tsconfig.json in parent directory
- Additional Dependencies needed for this script were all added as devDependencies in `package.json`

TODO's:
- update workflow file bot user as committer, token for bot user as well as path to bot user fork
- update workflow file with WIF provider and SA email account

bharathkkb commented on .github/workflows/updateReadme.yml:
```diff
@@ -0,0 +1,53 @@
+        node-version: '16.x'
+    
+    # Used to authenticate setup-gcloud in update script
+    - name: 'Auth with WIF'
+      uses: google-github-actions/auth@v0
+      with:
+        workload_identity_provider: # TODO replace with WIF
+        service_account: # TODO: replace with SA
```
Do we need auth for this? `docker run -ti  google/cloud-sdk:latest gcloud functions event-types list` seems to work without auth

LukaFontanilla replied:
Had not realized auth wouldn't be required (or rather had not tested without). Can omit the auth step and just run the command with a setup gcloud step prior.

bharathkkb commented on utils/updateReadme.ts:
```diff
@@ -0,0 +1,87 @@
+      if (err)
+        throw new Error(`There was an error reading the moustache file template. Error: ${err}`);
+      const output = Moustache.render(data.toString(), eventTriggerData);
+      fs.writeFileSync('../READMETEST.md', output);
```
nit: I think this was for testing?

bharathkkb commented on utils/updateReadme.ts:
```diff
@@ -0,0 +1,87 @@
+  try {
+    const cmd = ['functions', 'event-types', 'list', '--format', 'json'];
+    // Setup latest Gcloud Version
+    const gcloudVersion = await setupGcloud.getLatestGcloudSDKVersion();
+
+    if (!setupGcloud.isInstalled(gcloudVersion)) {
+      await setupGcloud.installGcloudSDK(gcloudVersion);
+    } else {
+      const toolPath = toolCache.find('gcloud', gcloudVersion);
+      core.addPath(path.join(toolPath, 'bin'));
+    }
+    await setupGcloud.authenticateGcloudSDK();
+    const authenticated = await setupGcloud.isAuthenticated();
+    if (!authenticated) {
+      throw new Error('Error authenticating the Cloud SDK.');
```
Cant we use setup-gcloud as a pre step in the job that calls this so all of this is already setup?

LukaFontanilla replied:
Will do!

bharathkkb commented on utils/updateReadme.ts:
```diff
@@ -0,0 +1,87 @@
+    const { output } = await setupGcloud.gcloudRun(cmd);
+    return {
+      output: output,
+      commandString: commandString,
```
Do we need commandString? gcloudRun(cmd) will throw an error with `commandString` for non zero exit code.

LukaFontanilla replied:
Ended up running gcloud command as a separate step and removed this code.

bharathkkb reviewed (commented):
Thanks for the PR @LukaFontanilla 
Some initial comments

sethvargo commented on utils/main.moustache:
```diff
@@ -0,0 +1,316 @@
+<!--
```
We don't actually need this for markdown files.

sethvargo commented on utils/main.moustache:
```diff
@@ -0,0 +1,316 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+-->
+# deploy-cloud-functions
```
I'm not in favor of moving this entire thing into a template file. It makes external contributions difficult. Instead, could we render this list to another file in `docs/` and then just link to it from the main README instead?

LukaFontanilla replied:
Yep, I definitely understand. Had not thought about the complications of using a template file. Will rework with the `docs/` approach instead.

sethvargo commented on utils/updateReadme.ts:
```diff
@@ -0,0 +1,87 @@
+    const eventTriggerData = {
+      events: Array.from(uniqueEvents),
+    };
+    // read moustache template and write new readme to be commited
+    fs.readFile('./main.moustache', (err, data) => {
```
+1. I also suggested potentially moving the event types into a separate file in `docs/` and just linking from the static readme.

LukaFontanilla replied:
Thanks for sharing a similar example @bharathkkb, I can see how the template file would lead to complications. I'll look into this approach and rework.

sethvargo commented on .github/workflows/updateReadme.yml:
```diff
@@ -0,0 +1,53 @@
+        workload_identity_provider: # TODO replace with WIF
+        service_account: # TODO: replace with SA
+    
+    - name: 'Build & Run Update Readme Script'
```
I might be missing something, but this seems a little complex to me. I think we could have three steps:

1. `uses: google-github-actions/setup-gcloud`
2. `run: gcloud --format=json functions event-types list > events.json`
3. render template with events.json as input

LukaFontanilla commented:
Implemented a similar approach to what y'all had mentioned: 

- `docs` file with event table markdown that script writes to (uses html comments to identify substring to be replaced in markdown file)
- `gcloud` command as step in workflow outputting results to an `events.json` file
- simplified the `updateReadme.ts` to account for this and removed all previously added dependencies
